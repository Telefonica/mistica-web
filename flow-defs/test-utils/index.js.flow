// @flow
/**
 * Flowtype definitions for index
 * Generated by Flowgen from a Typescript Definition
 * Flowgen v1.10.0
 */

import {
  type Page,
  type ElementHandle,
  type ClickOptions,
  type ScreenshotOptions
} from "puppeteer";
import { type Skin } from "../colors";
declare var MOBILE_DEVICE_IOS: "MOBILE_IOS";
declare var MOBILE_DEVICE_ANDROID: "MOBILE_ANDROID";
declare var TABLET_DEVICE: "TABLET";
declare var DESKTOP_DEVICE: "DESKTOP";
export type Device =
  | typeof MOBILE_DEVICE_IOS
  | typeof MOBILE_DEVICE_ANDROID
  | typeof TABLET_DEVICE
  | typeof DESKTOP_DEVICE;
export type PageApi = {
  type: (
    selector: ElementHandle | Promise<ElementHandle>,
    text: string,
    options?: {
      delay: number
    }
  ) => Promise<void>,
  click: (
    selector: ElementHandle | Promise<ElementHandle>,
    options?: ClickOptions
  ) => Promise<void>,
  screenshot: (
    options?: ScreenshotOptions
  ) => $Call<<R>((...args: any[]) => R) => R, $PropertyType<Page, "screenshot">>
};
declare type Query = (m: string) => Promise<ElementHandle>;
declare type AllQuery = (m: string) => Promise<Array<ElementHandle>>;
declare type Queries = {
  getByText: Query,
  getAllByText: AllQuery,
  getByTestId: Query,
  getAllByTestId: AllQuery,
  getByTitle: Query,
  getAllByTitle: AllQuery,
  getByRole: Query,
  getAllByRole: AllQuery,
  getByPlaceholderText: Query,
  getAllByPlaceholderText: AllQuery,
  getByLabelText: Query,
  getAllByLabelText: AllQuery,
  getByAltText: Query,
  getAllByAltText: AllQuery
};
declare export var openStoryPage: (x: {
  section: string,
  name: string,
  device?: "MOBILE_IOS" | "MOBILE_ANDROID" | "TABLET" | "DESKTOP" | void,
  skin?: "Movistar" | "O2" | "Vivo" | void
}) => Promise<PageApi>;
declare export var screen: Queries;
declare export {};
