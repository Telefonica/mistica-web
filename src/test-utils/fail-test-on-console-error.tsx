let errorOrWarnCalled = false;
const originalConsoleError = console.error;
const originalConsoleWarn = console.warn;

beforeEach(() => {
    jest.spyOn(global.console, 'error').mockImplementation((...args) => {
        errorOrWarnCalled = true;
        originalConsoleError(...args);
    });

    jest.spyOn(global.console, 'warn').mockImplementation((...args) => {
        errorOrWarnCalled = true;
        originalConsoleWarn(...args);
    });
});

afterEach(() => {
    if (errorOrWarnCalled) {
        throw new Error('Console warnings and errors are not allowed in tests.');
    }
});
