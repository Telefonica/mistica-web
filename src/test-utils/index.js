// @flow
import {getDocument, queries} from 'pptr-testing-library';
import jimp from 'jimp';

import type {Page, ElementHandle, ClickOptions} from 'puppeteer';
import type {Viewport} from 'puppeteer/DeviceDescriptors';
import type {Skin} from '../colors';

const STORYBOOK_URL = ((): string => {
    if (global.browser) {
        const url = new URL(global.browser.wsEndpoint());
        const isUsingDockerizedChromium = url.port === '9223';
        if (isUsingDockerizedChromium) {
            return process.platform === 'linux'
                ? 'http://172.17.0.1:6006/iframe.html'
                : 'http://host.docker.internal:6006/iframe.html';
        }
    }
    return 'http://localhost:6006/iframe.html';
})();

const MOBILE_DEVICE_IOS: 'MOBILE_IOS' = 'MOBILE_IOS';
const MOBILE_DEVICE_ANDROID: 'MOBILE_ANDROID' = 'MOBILE_ANDROID';
const TABLET_DEVICE: 'TABLET' = 'TABLET';
const DESKTOP_DEVICE: 'DESKTOP' = 'DESKTOP';

type Device =
    | typeof MOBILE_DEVICE_IOS
    | typeof MOBILE_DEVICE_ANDROID
    | typeof TABLET_DEVICE
    | typeof DESKTOP_DEVICE;

const DEVICES: {
    [name: Device]: {
        platform?: string,
        viewport: Viewport,
    },
    ...
} = {
    [MOBILE_DEVICE_IOS]: {
        platform: 'ios',
        viewport: {
            width: 375,
            height: 667,
            deviceScaleFactor: 2,
            isMobile: true,
            hasTouch: true,
            isLandscape: false,
        },
    },
    [MOBILE_DEVICE_ANDROID]: {
        platform: 'android',
        viewport: {
            width: 411,
            height: 731,
            deviceScaleFactor: 2,
            isMobile: true,
            hasTouch: true,
            isLandscape: false,
        },
    },
    [TABLET_DEVICE]: {
        // Use default platform
        platform: undefined,
        viewport: {
            width: 800,
            height: 600,
            deviceScaleFactor: 1,
            isMobile: false,
            hasTouch: false,
            isLandscape: false,
        },
    },
    [DESKTOP_DEVICE]: {
        // Use default platform
        platform: undefined,
        viewport: {
            width: 1280,
            height: 800,
            deviceScaleFactor: 1,
            isMobile: false,
            hasTouch: false,
            isLandscape: false,
        },
    },
};

const STEP_TIME = 250;

type WaitForPaintEndOptions = {
    maxWait?: number,
    fullPage?: boolean,
};

const waitForPaintEnd = async (element, {maxWait = 10000, fullPage = true}: WaitForPaintEndOptions = {}) => {
    const t0 = Date.now();

    let buf1 = await element.screenshot({fullPage});
    await new Promise((r) => setTimeout(r, STEP_TIME));
    let buf2 = await element.screenshot({fullPage});

    // buffers are different if compare != 0
    while (buf1.compare(buf2)) {
        if (Date.now() - t0 > maxWait) {
            throw Error('Paint end timeout');
        }
        buf1 = buf2;
        await new Promise((r) => setTimeout(r, STEP_TIME));
        buf2 = await element.screenshot({fullPage});
    }
};

const watermarkIfNeeded = async (bufferPromise: Promise<Buffer>): Promise<Buffer> => {
    if (process.env.HEADLESS || process.env.CI) {
        return bufferPromise;
    }
    const image = await jimp.read(await bufferPromise);
    image.color([
        {apply: 'desaturate', params: [100]},
        {apply: 'tint', params: [50]},
    ]);
    await jimp.loadFont(jimp.FONT_SANS_32_BLACK).then((font) => {
        image.print(
            font,
            0,
            0,
            {
                text: 'Screenshot not valid for CI. Please, regenerate in headless mode',
                alignmentX: jimp.HORIZONTAL_ALIGN_CENTER,
                alignmentY: jimp.VERTICAL_ALIGN_MIDDLE,
            },
            image.bitmap.width,
            image.bitmap.height
        );
    });
    return image.getBufferAsync(jimp.MIME_PNG);
};

const buildStoryUrl = (section: string, name: string, skin: ?string, platform: ?string) => {
    const params = new URLSearchParams();
    params.set('selectedKind', section);
    params.set('selectedStory', name);
    if (skin) {
        params.set('skin', skin);
    }
    if (platform) {
        params.set('platform', platform);
    }
    return `${STORYBOOK_URL}?${params.toString()}`;
};

type PageApi = {
    // Following methods are inherited from Puppeteer.Page, some are overridden

    // These are overridden:
    type: (
        selector: ElementHandle | Promise<ElementHandle>,
        text: string,
        options?: {delay: number}
    ) => Promise<void>,
    click: (selector: ElementHandle | Promise<ElementHandle>, options?: ClickOptions) => Promise<void>,
    // We need to use this convoluted $Call because $PropertyType does not work with Interfaces, and Page is
    // declared as an Interface, not an Object type

    // These are from prototype chain (inherited from Puppeteer.Page)
    screenshot: $Call<<T, P: {+screenshot: T, ...}>(P) => T, Page>,
};

const wait = <T>(
    expectation: () => Promise<T> | T,
    timeout?: number = 4500,
    interval?: number = 50
): Promise<T> => {
    const startTime = Date.now();
    return new Promise((resolve, reject) => {
        const rejectOrRerun = (error) => {
            if (Date.now() - startTime >= timeout) {
                reject(error);
                return;
            }
            // eslint-disable-next-line no-use-before-define
            setTimeout(runExpectation, interval);
        };
        const runExpectation = () => {
            try {
                Promise.resolve(expectation())
                    .then((r) => resolve(r))
                    .catch(rejectOrRerun);
            } catch (error) {
                rejectOrRerun(error);
            }
        };
        setTimeout(runExpectation, 0);
    });
};

const bindToDoc = (fn) => async (m: string) => {
    const doc = await getDocument(global.page);
    const body = await doc.$('body');
    const elementHandle = await wait(() => fn(body, m));

    const screenshot = async (options) => {
        await waitForPaintEnd(elementHandle, {fullPage: false});
        return watermarkIfNeeded(elementHandle.screenshot(options));
    };

    const newElementHandle = Object.create(elementHandle);
    newElementHandle.screenshot = screenshot;
    return newElementHandle;
};

type Query = (m: string) => Promise<ElementHandle>;
type AllQuery = (m: string) => Promise<Array<ElementHandle>>;

type Queries = {
    getByText: Query,
    getAllByText: AllQuery,
    getByTestId: Query,
    getAllByTestId: AllQuery,
    getByTitle: Query,
    getAllByTitle: AllQuery,
    getByRole: Query,
    getAllByRole: AllQuery,
    getByPlaceholderText: Query,
    getAllByPlaceholderText: AllQuery,
    getByLabelText: Query,
    getAllByLabelText: AllQuery,
    getByAltText: Query,
    getAllByAltText: AllQuery,
};

const buildQueryMethods = (): any => {
    return Object.entries(queries).reduce(
        (bindedQueries, [queryName: $Keys<Queries>, queryFn]) => ({
            ...bindedQueries,
            [queryName]: bindToDoc(queryFn),
        }),
        {}
    );
};

const createPageApi = (page: Page): PageApi => {
    // $FlowFixMe I know, flow. This is a bit hacky
    const api: PageApi = Object.create(page);

    api.type = async (selector, text, options) => (await selector).type(text, options);
    api.click = async (selector, options) => (await selector).click(options);
    api.screenshot = async (options) => {
        await waitForPaintEnd(page);
        return watermarkIfNeeded(page.screenshot(options));
    };

    return api;
};

export const openStoryPage = async ({
    section,
    name,
    device = TABLET_DEVICE,
    skin = 'Movistar',
}: {
    section: string,
    name: string,
    device?: Device,
    skin?: Skin,
}): Promise<PageApi> => {
    const page: Page = global.page;
    const browser = global.browser;
    await page.bringToFront();
    await page.setViewport(DEVICES[device].viewport);
    await page.setUserAgent(`${await browser.userAgent()} acceptance-test`);
    await page.goto(buildStoryUrl(section, name, skin, DEVICES[device].platform));

    return createPageApi(page);
};

export const screen: Queries = buildQueryMethods();
