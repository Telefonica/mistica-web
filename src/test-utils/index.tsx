import {
    openPage,
    serverHostName,
    screen,
    getGlobalPage,
    PageApi,
    Viewport,
} from '@telefonica/acceptance-testing';
import {MOVISTAR_SKIN} from '../skins/constants';

export type {ElementHandle} from '@telefonica/acceptance-testing';
export {screen, PageApi};

export type StoryArgs = {[key: string]: string | number | boolean};

const MOBILE_DEVICE_IOS_SMALL = 'MOBILE_IOS_SMALL';
const MOBILE_DEVICE_IOS = 'MOBILE_IOS';
const MOBILE_DEVICE_ANDROID = 'MOBILE_ANDROID';
const TABLET_DEVICE = 'TABLET';
const DESKTOP_DEVICE = 'DESKTOP';

export type Device =
    | typeof MOBILE_DEVICE_IOS_SMALL
    | typeof MOBILE_DEVICE_IOS
    | typeof MOBILE_DEVICE_ANDROID
    | typeof TABLET_DEVICE
    | typeof DESKTOP_DEVICE;

type DeviceCollection = Record<
    Device,
    {
        platform?: string;
        userAgent?: string;
        viewport: Viewport;
    }
>;

const DEVICES: DeviceCollection = {
    [MOBILE_DEVICE_IOS_SMALL]: {
        userAgent:
            'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1',
        platform: 'ios',
        viewport: {
            width: 320,
            height: 480,
            deviceScaleFactor: 2,
            isMobile: true,
            hasTouch: true,
            isLandscape: false,
        },
    },
    [MOBILE_DEVICE_IOS]: {
        userAgent:
            'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1',
        platform: 'ios',
        viewport: {
            width: 375,
            height: 667,
            deviceScaleFactor: 2,
            isMobile: true,
            hasTouch: true,
            isLandscape: false,
        },
    },
    [MOBILE_DEVICE_ANDROID]: {
        userAgent:
            'Mozilla/5.0 (Linux; Android 8.0; Pixel 2 Build/OPD3.170816.012) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/79.0.3945.130 Mobile Safari/537.36',
        platform: 'android',
        viewport: {
            width: 411,
            height: 731,
            deviceScaleFactor: 2,
            isMobile: true,
            hasTouch: true,
            isLandscape: false,
        },
    },
    [TABLET_DEVICE]: {
        // Use default platform
        platform: undefined,
        viewport: {
            width: 800,
            height: 600,
            deviceScaleFactor: 1,
            isMobile: false,
            hasTouch: false,
            isLandscape: false,
        },
        userAgent: '',
    },
    [DESKTOP_DEVICE]: {
        // Use default platform
        platform: undefined,
        viewport: {
            width: 1280,
            height: 800,
            deviceScaleFactor: 1,
            isMobile: false,
            hasTouch: false,
            isLandscape: false,
        },
    },
};

const buildStoryPath = (id: string, skin?: string, platform?: string, args?: StoryArgs) => {
    const params = new URLSearchParams();

    params.set('id', id);

    params.set('viewMode', 'story');

    if (skin) {
        params.set('skin', skin);
    }

    if (platform) {
        params.set('platform', platform);
    }

    if (args) {
        params.set(
            'args',
            Object.entries(args)
                .map(([key, value]) => `${key}:${value}`)
                .join(';')
        );
    }

    return `/iframe.html?${params.toString()}`;
};

export const openStoryPage = ({
    id,
    device = TABLET_DEVICE,
    viewport,
    skin = 'Movistar',
    args,
    isDarkMode,
}: {
    id: string;
    device?: Device;
    viewport?: Viewport;
    skin?: 'Movistar' | 'Vivo' | 'O2' | 'O2-classic' | 'Blau';
    args?: StoryArgs;
    isDarkMode?: boolean;
}): Promise<PageApi> =>
    openPage({
        path: buildStoryPath(id, skin, DEVICES[device].platform, args),
        userAgent: DEVICES[device].userAgent,
        viewport: viewport ?? DEVICES[device].viewport,
        isDarkMode,
    });

/**
 * Renders a page with a React component in the server and opens it in the browser, where it's hydrated client side.
 * `name` is the name (without extension) of a file in the __ssr_pages__ folder. This file exports the component to be rendered.
 */
export const openSSRPage = ({
    name,
    device = TABLET_DEVICE,
    skin = MOVISTAR_SKIN,
}: {
    name: string;
    device?: Device;
    skin?: string;
}): Promise<PageApi> => {
    const page = getGlobalPage();
    const port = (global as any)['__SSR_SERVER__'].address().port;

    // Capture browser console.error and console.warn calls that React could trigger when calling hydrate()
    page.on('console', async (msg) => {
        const type = msg.type();
        const args = [...(await Promise.all(msg.args().map((h: any) => h.jsonValue())))];
        if (args.length === 0) {
            args.push(msg.text());
        }
        if (type === 'error') {
            console.error(...args);
        }
        if (type === 'warning') {
            console.warn(...args);
        }
    });

    return openPage({
        url: `http://${serverHostName}:${port}/${name}?skin=${skin}`,
        userAgent: DEVICES[device].userAgent,
        viewport: DEVICES[device].viewport,
    });
};

export const setRootFontSize = (px: number): Promise<void> =>
    page.$eval('html', (e, px) => e.setAttribute('style', `font-size: ${px}px;`), px);

export const waitFor = <T,>(
    expectation: () => Promise<T> | T,
    timeout = 10000,
    interval = 50
): Promise<T> => {
    const startTime = Date.now();
    return new Promise((resolve, reject) => {
        const rejectOrRerun = (error: unknown) => {
            if (Date.now() - startTime >= timeout) {
                reject(error);
                return;
            }
            // eslint-disable-next-line @typescript-eslint/no-use-before-define
            setTimeout(runExpectation, interval);
        };
        const runExpectation = () => {
            try {
                Promise.resolve(expectation())
                    .then((r) => resolve(r))
                    .catch(rejectOrRerun);
            } catch (error: any) {
                rejectOrRerun(error);
            }
        };
        setTimeout(runExpectation, 0);
    });
};
