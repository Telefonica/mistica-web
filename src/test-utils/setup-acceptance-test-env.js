// @flow
import {configureToMatchImageSnapshot} from 'jest-image-snapshot';

const toMatchImageSnapshot = configureToMatchImageSnapshot({
    failureThreshold: '0.0005',
    failureThresholdType: 'percent',
});

expect.extend({toMatchImageSnapshot});
