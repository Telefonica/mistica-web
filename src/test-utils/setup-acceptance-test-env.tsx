import {configureToMatchImageSnapshot} from 'jest-image-snapshot';

const toMatchImageSnapshot = configureToMatchImageSnapshot({
    failureThreshold: 0.0005,
    failureThresholdType: 'percent',
});

const argv = process.execArgv.join();
const isDebug = argv.includes('inspect') || argv.includes('debug');
jest.setTimeout(isDebug ? 3600000 : 30000);

expect.extend({toMatchImageSnapshot});
