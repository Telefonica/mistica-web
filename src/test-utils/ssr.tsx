import * as React from 'react';
import * as url from 'url';
import ReactDomServer from 'react-dom/server';
import path from 'path';
import webpack from 'webpack';
import http from 'http';
import fs from 'fs';
import {ThemeContextProvider, MOVISTAR_SKIN, getSkinByName, type KnownSkinName} from '../..';
import {execSync} from 'child_process';

const createWebpackEntries = (): {[entryName: string]: string} => {
    const entries: {[entryName: string]: string} = {};

    // create an entry for each .tsx file in __ssr_pages__ folder
    const files = fs
        .readdirSync(path.join(__dirname, '..', '__acceptance_tests__/__ssr_pages__'))
        .filter((f) => f.endsWith('.tsx'));

    files.forEach((f) => {
        const moduleName = f.replace('.tsx', ''); // remove file extension

        // create a temporary file for the entry, it'll be removed after webpack compilation
        const clientSideEntryFile = path.join(__dirname, `temporary-file-do-not-commit-${f}`);
        fs.writeFileSync(
            clientSideEntryFile,
            `
            import * as React from 'react';
            import { hydrateRoot } from 'react-dom/client';
            import Component from '../__acceptance_tests__/__ssr_pages__/${moduleName}';
            import {ThemeContextProvider, getSkinByName} from '../..';

            const skin = new URL(location).searchParams.get('skin');

            hydrateRoot(
                document.getElementById('root'),
                <ThemeContextProvider
                    theme={{
                        skin: getSkinByName(skin || 'Movistar'),
                        i18n: {locale: 'es-ES', phoneNumberFormattingRegionCode: 'ES'}
                    }}
                >
                    <Component />
                </ThemeContextProvider>
            );`
        );

        entries[moduleName] = clientSideEntryFile;
    });

    return entries;
};

export const compileSsrClient = ({build = true}: {build: boolean}): Promise<webpack.Stats> => {
    if (build) {
        execSync('yarn compile', {stdio: 'inherit'});
    }
    const entries = createWebpackEntries();
    const webpackConfig: webpack.Configuration = {
        mode: 'development',
        // While testing NODE_ENV will get evaluated to test on the server, so it should be the same on the browser bundles.
        optimization: {
            nodeEnv: 'test',
        },
        entry: entries,
        output: {
            path: path.resolve(__dirname, '..', '..', 'public', 'ssr'),
            filename: '[name].js',
            publicPath: '',
        },
        module: {
            rules: [
                {
                    test: /\.tsx?$/,
                    use: {
                        loader: 'swc-loader',
                    },
                    exclude: [/node_modules/],
                },
            ],
        },
        resolve: {
            extensions: ['.tsx', '.ts', '.js', '.json', '.wasm', '.mjs', '*'],
        },
        plugins: [
            new webpack.DefinePlugin({
                'process.env.SSR_TEST': JSON.stringify(true),
            }),
        ],
    };

    const stats = new Promise<webpack.Stats>((resolve, reject) => {
        const compiler = webpack(webpackConfig);
        compiler.run((err, stats: any) => {
            if (err || stats.hasErrors()) {
                reject(stats.compilation.errors);
            }
            resolve(stats);
        });
    }).finally(() => {
        // remove all temporary entry files
        Object.values(entries).forEach((file) => {
            fs.unlinkSync(file);
        });
    });

    return stats;
};

export const createServer = (): http.Server => {
    const server = http.createServer((req, res) => {
        const parsedUrl = url.parse(req.url || '', true);
        const pathParts = (parsedUrl.pathname || '').split('/');
        const moduleName = pathParts[pathParts.length - 1];

        if (moduleName === 'favicon.ico') {
            res.end();
            return;
        }

        const serveFileInPath = (path: string) => {
            fs.readFile(path, (err, data) => {
                if (err) {
                    throw err;
                }
                res.writeHead(200);
                res.end(data);
            });
        };

        if (moduleName.endsWith('.woff2')) {
            serveFileInPath(path.join(__dirname, '..', '..', '.storybook', 'css', parsedUrl.path as string));
            return;
        }

        if (moduleName.endsWith('.css')) {
            serveFileInPath(path.join(__dirname, '..', '..', 'css', parsedUrl.path as string));
            return;
        }

        let Component;
        try {
            // eslint-disable-next-line @typescript-eslint/no-var-requires
            Component = require('../__acceptance_tests__/__ssr_pages__/' + moduleName).default;
        } catch (e) {
            res.writeHead(404);
            res.end();
            return;
        }

        const userAgent = req.headers['user-agent'];

        const renderedComponent = ReactDomServer.renderToString(
            <ThemeContextProvider
                theme={{
                    skin: getSkinByName(String(parsedUrl.query.skin || MOVISTAR_SKIN) as KnownSkinName),
                    i18n: {locale: 'es-ES', phoneNumberFormattingRegionCode: 'ES'},
                    platformOverrides: {
                        userAgent,
                    },
                }}
            >
                <Component />
            </ThemeContextProvider>
        );

        const clientCode = fs.readFileSync(
            path.join(__dirname, '..', '..', 'public', 'ssr', `${moduleName}.js`)
        );

        res.writeHead(200);

        res.end(`
            <!DOCTYPE html>
            <html>
                <head>
                    <meta charset="UTF-8">
                    <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0,minimum-scale=1.0,user-scalable=no,viewport-fit=cover">
                    <link rel="stylesheet" href="reset.css">
                    <link rel="stylesheet" href="roboto.css">
                    <link rel="stylesheet" href="mistica.css">
                </head>
                <body>
                    <div id="root">${renderedComponent}</div>
                    <script>${clientCode}</script>
                </body>
            </html>`);
    });

    return server;
};
