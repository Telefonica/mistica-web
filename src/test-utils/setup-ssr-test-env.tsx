import './fail-test-on-console-error';
import {createServer} from './ssr';

process.env.SSR_TEST = 'true';

beforeAll(() => {
    (global as any)['__SSR_SERVER__'] = createServer();
    (global as any)['__SSR_SERVER__'].listen(0); // use any free port
});

afterAll(() => {
    (global as any)['__SSR_SERVER__'].close();
});
