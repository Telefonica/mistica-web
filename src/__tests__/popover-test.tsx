import * as React from 'react';
import {fireEvent, render, waitFor, screen} from '@testing-library/react';
import Popover from '../popover';
import {ThemeContextProvider} from '..';
import {makeTheme} from './test-utils';

type Props = Omit<React.ComponentProps<typeof Popover>, 'description' | 'target'>;

const TestPopover: React.FC<Props> = ({children, ...props}) => (
    <ThemeContextProvider theme={makeTheme()}>
        <Popover {...props} description="Content" target={<span>Press me!</span>} />
    </ThemeContextProvider>
);

const TestWrapper = ({onCloseSpy}: {onCloseSpy: () => void}) => {
    const [isVisible, setIsVisible] = React.useState(true);
    const handleClose = () => {
        onCloseSpy();
        setIsVisible(false);
    };
    return (
        <ThemeContextProvider theme={makeTheme()}>
            <Popover
                isVisible={isVisible}
                description="Content"
                onClose={handleClose}
                target={<span>Press me!</span>}
            />
        </ThemeContextProvider>
    );
};

const AnyIcon = () => <svg data-testid="icon" />;

test('target is painted', () => {
    render(<TestPopover />);

    expect(screen.getByText('Press me!')).toBeInTheDocument();
});

test('popover is painted by default', () => {
    render(<TestPopover />);

    expect(screen.getByText('Content')).toBeInTheDocument();
    expect(screen.getByLabelText('Cerrar')).toBeInTheDocument();
});

test('popover is not painted if visible property is false', () => {
    render(<TestPopover isVisible={false} />);

    expect(screen.queryByText('Content')).toBe(null);
});

test('set title and description', () => {
    render(<TestPopover title="Title" />);

    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByText('Content')).toBeInTheDocument();
});

test('set title and asset', () => {
    render(<TestPopover title="Title" asset={<AnyIcon />} />);

    expect(screen.getByText('Title')).toBeInTheDocument();
    expect(screen.getByTestId('icon')).toBeInTheDocument();
});

test('popover is not painted after click the close icon', async () => {
    const onCloseSpy = jest.fn();
    render(<TestWrapper onCloseSpy={onCloseSpy} />);
    expect(screen.getByText('Content')).toBeInTheDocument();
    fireEvent.click(screen.getByLabelText('Cerrar'));

    await waitFor(() => expect(onCloseSpy).toHaveBeenCalled());

    expect(screen.queryByText('Content')).toBe(null);
});
