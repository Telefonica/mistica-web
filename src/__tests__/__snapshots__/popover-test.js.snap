// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popover default(bottom) style 1`] = `
<div>
  <div
    class="wrapper-0-2-3"
  >
    <div>
      <span>
        Press me!
      </span>
    </div>
    <div
      class="container-0-2-4"
      style="width: 488px; top: 20px; left: 0px;"
    >
      <div
        class="arrowWrapper-0-2-2"
        style="bottom: 100%; left: 0px; transform: translateX(-50%) rotate(180deg);"
      >
        <div
          class="arrow-0-2-1 arrow-d6-0-2-19"
        />
      </div>
      <div
        class="boxContainer-0-2-6"
      >
        <div
          class="textContent-0-2-7"
        >
          <span
            class="text-0-2-9"
          >
            Content
          </span>
        </div>
        <div
          class="closeButtonIcon-0-2-10"
        >
          <button
            aria-label="Cerrar"
            class="touchable-0-2-12"
            style="display: inline-block; vertical-align: middle; text-align: center; background-color: transparent; background-position: 50% 50%; background-size: 100% 100%; background-repeat: no-repeat; border: 0px; cursor: pointer; height: 24px; width: 24px;"
            type="button"
          >
            <svg
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M17.4 6.6a.849.849 0 0 0-1.2 0L12 10.8 7.8 6.6a.849.849 0 1 0-1.2 1.2l4.2 4.2-4.2 4.2a.849.849 0 0 0 1.2 1.2l4.2-4.2 4.2 4.2a.849.849 0 0 0 1.2-1.2L13.2 12l4.2-4.2a.849.849 0 0 0 0-1.2z"
                fill="#313235"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`popover left style 1`] = `
<div>
  <div
    class="wrapper-0-2-3"
  >
    <div>
      <span>
        Press me!
      </span>
    </div>
    <div
      class="container-0-2-4"
      style="width: 488px; left: -508px; top: 0px; transform: translateY(-50%);"
    >
      <div
        class="arrowWrapper-0-2-2"
        style="left: 100%; top: 50%; transform: translateY(-100%) rotate(-90deg); transform-origin: bottom;"
      >
        <div
          class="arrow-0-2-1 arrow-d9-0-2-22"
        />
      </div>
      <div
        class="boxContainer-0-2-6"
      >
        <div
          class="textContent-0-2-7"
        >
          <span
            class="text-0-2-9"
          >
            Content
          </span>
        </div>
        <div
          class="closeButtonIcon-0-2-10"
        >
          <button
            aria-label="Cerrar"
            class="touchable-0-2-12"
            style="display: inline-block; vertical-align: middle; text-align: center; background-color: transparent; background-position: 50% 50%; background-size: 100% 100%; background-repeat: no-repeat; border: 0px; cursor: pointer; height: 24px; width: 24px;"
            type="button"
          >
            <svg
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M17.4 6.6a.849.849 0 0 0-1.2 0L12 10.8 7.8 6.6a.849.849 0 1 0-1.2 1.2l4.2 4.2-4.2 4.2a.849.849 0 0 0 1.2 1.2l4.2-4.2 4.2 4.2a.849.849 0 0 0 1.2-1.2L13.2 12l4.2-4.2a.849.849 0 0 0 0-1.2z"
                fill="#313235"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`popover right style 1`] = `
<div>
  <div
    class="wrapper-0-2-3"
  >
    <div>
      <span>
        Press me!
      </span>
    </div>
    <div
      class="container-0-2-4"
      style="width: 488px; left: 20px; top: 0px; transform: translateY(-50%);"
    >
      <div
        class="arrowWrapper-0-2-2"
        style="right: 100%; top: 50%; transform: translateY(-100%) rotate(90deg); transform-origin: bottom;"
      >
        <div
          class="arrow-0-2-1 arrow-d8-0-2-21"
        />
      </div>
      <div
        class="boxContainer-0-2-6"
      >
        <div
          class="textContent-0-2-7"
        >
          <span
            class="text-0-2-9"
          >
            Content
          </span>
        </div>
        <div
          class="closeButtonIcon-0-2-10"
        >
          <button
            aria-label="Cerrar"
            class="touchable-0-2-12"
            style="display: inline-block; vertical-align: middle; text-align: center; background-color: transparent; background-position: 50% 50%; background-size: 100% 100%; background-repeat: no-repeat; border: 0px; cursor: pointer; height: 24px; width: 24px;"
            type="button"
          >
            <svg
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M17.4 6.6a.849.849 0 0 0-1.2 0L12 10.8 7.8 6.6a.849.849 0 1 0-1.2 1.2l4.2 4.2-4.2 4.2a.849.849 0 0 0 1.2 1.2l4.2-4.2 4.2 4.2a.849.849 0 0 0 1.2-1.2L13.2 12l4.2-4.2a.849.849 0 0 0 0-1.2z"
                fill="#313235"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`popover top style 1`] = `
<div>
  <div
    class="wrapper-0-2-3"
  >
    <div>
      <span>
        Press me!
      </span>
    </div>
    <div
      class="container-0-2-4"
      style="width: 488px; transform: translateY(-100%); top: -20px; left: 0px;"
    >
      <div
        class="arrowWrapper-0-2-2"
        style="top: 100%; left: 0px; transform: translateX(-50%);"
      >
        <div
          class="arrow-0-2-1 arrow-d7-0-2-20"
        />
      </div>
      <div
        class="boxContainer-0-2-6"
      >
        <div
          class="textContent-0-2-7"
        >
          <span
            class="text-0-2-9"
          >
            Content
          </span>
        </div>
        <div
          class="closeButtonIcon-0-2-10"
        >
          <button
            aria-label="Cerrar"
            class="touchable-0-2-12"
            style="display: inline-block; vertical-align: middle; text-align: center; background-color: transparent; background-position: 50% 50%; background-size: 100% 100%; background-repeat: no-repeat; border: 0px; cursor: pointer; height: 24px; width: 24px;"
            type="button"
          >
            <svg
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
            >
              <path
                d="M17.4 6.6a.849.849 0 0 0-1.2 0L12 10.8 7.8 6.6a.849.849 0 1 0-1.2 1.2l4.2 4.2-4.2 4.2a.849.849 0 0 0 1.2 1.2l4.2-4.2 4.2 4.2a.849.849 0 0 0 1.2-1.2L13.2 12l4.2-4.2a.849.849 0 0 0 0-1.2z"
                fill="#313235"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
